projection/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -507,43 +544,57 @@ private[projection] class R2dbcOffsetStore(
 
     if (duplicate) {
       logger.debug("Filtering out duplicate sequence number [{}] for pid [{}]", seqNr, pid) // FIXME change to trace
-      false
+      FutureFalse
     } else if (recordWithOffset.strictSeqNr) {
       // strictSeqNr == true is for event sourced
       val prevSeqNr = currentInflight.getOrElse(pid, currentState.byPid.get(pid).map(_.seqNr).getOrElse(0L))
       if (prevSeqNr > 0) {
         // expecting seqNr to be +1 of previously known
         val ok = seqNr == prevSeqNr + 1
-        if (!ok)
-          logger.debug(
-            "Filtering out unexpected sequence number [{}] for pid [{}], previous sequence number [{}]",
+        if (ok) {
+          FutureTrue
+        } else {
+          logger.info(
+            "Rejecting unexpected sequence number [{}] for pid [{}], previous sequence number [{}]",
             seqNr,
             pid,
             prevSeqNr)
-        ok
+          FutureFalse
+        }
       } else if (seqNr == 1) {
         // always accept first event if no other event for that pid has been seen
-        true
+        FutureTrue
       } else {
         // Haven't see seen this pid within the time window. Since events can be missed
-        // when read at the tail we will only accept it if read after the acceptNewSequenceNumberAfterAge
-        // duration. Backtracking will emit it again.
-        val ok = JDuration
-          .between(timestampOffset.timestamp, timestampOffset.readTimestamp)
-          .compareTo(settings.acceptNewSequenceNumberAfterAge) >= 0
-        if (!ok)
-          logger.debug("Filtering out unknown sequence number (might be accepted later): {}", recordWithOffset)
-        ok
+        // when read at the tail we will only accept it if the event with previous seqNr has timestamp
+        // before the time window of the offset store.
+        // Backtracking will emit missed event again.
+        val entityTypeHint = SliceUtils.extractEntityTypeHintFromPersistenceId(pid)
+        val slice = SliceUtils.sliceForPersistenceId(pid, maxNumberOfSlices)
+        eventTimestampQuery.timestampOf(entityTypeHint, pid, slice, seqNr - 1).map {
+          case Some(previousTimestamp) =>
+            if (previousTimestamp.isBefore(currentState.latestTimestamp.minus(settings.timeWindow)))
+              true
+            else {
+              logger.info("Rejecting unknown sequence number (might be accepted later): {}", recordWithOffset)
+              false
+            }
+          case None =>
+            // previous not found, could have been deleted
+            true
+        }
```

Comment:
Not important but map could perhaps be on parasitic ec

Reply:
I think it's doing too much (risk for introducing something that could throw exception in the future), and this case shouldn't be used often so not a hot perf path